feat(ec2): lookup security group by name

jumic opened the pull request:
Support looking up a security group by name.

Currently, looking up a security group is only possible by ID. This PR enhances the existing implementation to support lookup by security group name.

`securityGroupName` or `securityGroupId` can be passed to the new method `SecurityGroup.fromLookupAttributes`. In addition, property `vpc` provides the option to restrict the lookup method to a specific VPC.

If no or more than one security group is found, an error is thrown.

Closes #4241.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 463e59935444b74b6c8c2f6ede7e4923a9baa5a5
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=54732a62-76b2-4da9-8940-afe5d7564077%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jumic commented:
The first build failed because property securityGroupId was changed to optional:
```
@aws-cdk/cloud-assembly-schema... CHANGES.
Original assembly: @aws-cdk/cloud-assembly-schema@1.130.0
Updated assembly:  @aws-cdk/cloud-assembly-schema@0.0.0
API elements with incompatible changes:
err  - IFACE @aws-cdk/cloud-assembly-schema.SecurityGroupContextQuery: formerly required property 'securityGroupId' is optional: type of property props, type of property missing, returned from @aws-cdk/cloud-assembly-schema.Manifest.load (...and 1 more...) [weakened:@aws-cdk/cloud-assembly-schema.SecurityGroupContextQuery]
```

I added this line to the breaking changes to allow this change.
```
weakened:@aws-cdk/cloud-assembly-schema.SecurityGroupContextQuery
```

Please check if this change is allowed.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 927ba53b70500e725ffe7b2f23f8af57f1a62a90
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1d4548c6-a78a-41ff-88a8-442e2c1b6343%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

fixundfertig123 commented:
@jumic sorry for asking this question but I am not that familiar with the approval/pipeline process of cdk. I just by coincident bumbed into this issue today when setting up a multi account pipeline. Therefore I would really appreciate if you could give me an indication when this feature hits GA. Would it be deployed to next release? Best, and thank you for your work!

rix0rrr commented on packages/@aws-cdk/aws-ec2/README.md:
````diff
@@ -577,6 +577,33 @@ const mySecurityGroupWithoutInlineRules = new ec2.SecurityGroup(this, 'SecurityG
+If you know the ID and the configuration of the security group to import, you can use `SecurityGroup.fromSecurityGroupId`:
+
+```ts
+const sg = ec2.SecurityGroup.fromSecurityGroupId(this, 'SecurityGroupImport', 'security-group-import', {
````
```suggestion
const sg = ec2.SecurityGroup.fromSecurityGroupId(this, 'SecurityGroupImport', 'sg-1234', {
```

njlynch commented on packages/@aws-cdk/aws-ec2/README.md:
````diff
@@ -577,6 +577,33 @@ const mySecurityGroupWithoutInlineRules = new ec2.SecurityGroup(this, 'SecurityG
+The result of the `SecurityGroup.fromLookupAttributes` operation will be written to a file called `cdk.context.json`. You must commit this file to source control so that the lookup values are available in non-privileged environments such as CI build steps, and to ensure your template builds are repeatable.
+
+```ts fixture=with-vpc
+const sg = ec2.SecurityGroup.fromLookupAttributes(
````
What do you think about dedicated `fromLookupById` and `fromLookupByName` methods? They could both still use a (private) `fromLookupAttributes` method, but it may make the API intent a bit more clear.

jumic replied:
I tried to create an API similar to [VpcLookupOptions](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.VpcLookupOptions.html). Now I compared it to `fromLookupById` and `fromLookupByName` again. For me, this seems to be a better option because the necessary parameters are defined by the API. I've implemented the following changes:
- Deprecated method `fromLookup`, will be replaced by `fromLookupById` (same parameters).
- New method `fromLookupById` with parameter `securityGroupId`. Parameter `vpc` not defined because the security group id is sufficient to clearly identify the security group.
- New method `fromLookupByName` with parameters `securityGroupName` and `vpc`.  `vpc` is defined as required parameter because it avoids problems if the same security group name is used multiple times per account and region.   A security group is used in context of a VPC. Therefore, the VPC should be available in the CDK code. If it should be defined as optional later, we could changed it easily (no breaking change).

njlynch commented on packages/@aws-cdk/aws-ec2/lib/security-group-lookup.ts:
```diff
@@ -0,0 +1,35 @@
+import { IVpc } from '.';
```
```suggestion
import { IVpc } from './vpc';
```

njlynch commented on packages/@aws-cdk/aws-ec2/lib/security-group.ts:
```diff
@@ -318,13 +319,32 @@ export class SecurityGroup extends SecurityGroupBase {
       throw new Error('All arguments to look up a security group must be concrete (no Tokens)');
     }
 
+    if (options.securityGroupId && options.securityGroupName) {
```
Having dedicated methods would avoid needing these checks.

njlynch reviewed (dismissed):
Looks mostly good. One comment about the public API for `SecurityGroup` I think is worth discussing.

jumic commented on packages/aws-cdk/lib/context-providers/security-groups.ts:
```diff
@@ -12,18 +12,45 @@ export class SecurityGroupContextProviderPlugin implements ContextProviderPlugin
     const account: string = args.account!;
     const region: string = args.region!;
 
+    if (args.securityGroupId && args.securityGroupName) {
+      throw new Error('\'securityGroupId\' and \'securityGroupName\' can not be specified both when looking up a security group');
+    }
+
+    if (!args.securityGroupId && !args.securityGroupName) {
+      throw new Error('\'securityGroupId\' or \'securityGroupName\' must be specified to look up a security group');
+    }
```
Should we remove or keep those checks? In module `ec2` I removed them because no public API allows to pass both or none of these parameters.

I would propose to keep the checks in this module. If somebody implements a custom construct using the logic from the context-providers, it would be helpful if these checks are available.

njlynch replied:
Agreed; worth keeping here down at this level.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f24916e4536f3ac653b9c5872f3264a150d1934d
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5e570586-1a2e-4431-95bf-4f89c1091a19%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 379dac81aca20817ad1ec678804767b8ae78efe1
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=44e04a18-20e2-4a10-8637-ebac67c2d51d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

njlynch reviewed (changes requested):
Looks great! One nitpicky code organization comment; otherwise good to go!

jumic commented on packages/@aws-cdk/aws-ec2/lib/security-group-lookup.ts:
```diff
@@ -0,0 +1,35 @@
+ *
+ * Either `securityGroupName` or `securityGroupId` has to be specified.
+ */
+export interface SecurityGroupLookupOptions {
```
Done. I moved it to the correct location.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a1ca5d8c7ca625a67f701cb6ba549a12572e2f45
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6fab0230-86c1-4509-84fb-bc8b5f2265c8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: aca049c27f11ea7464e92712685a6a00d2a0e465
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c4ef1d44-48db-4ab2-be02-587ec7094586%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

njlynch reviewed (approved):
🎉

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 6f85178c8f99031b391eb9d07d9cd9a009f4e1b1
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3b7a35ea-e6a4-4ab0-b703-bda0f2b7c5d0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
